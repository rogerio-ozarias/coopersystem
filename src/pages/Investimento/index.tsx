import React, { useEffect, useState } from 'react';
import api from "../../services/api";
import { formatNumber } from "../../helpers/Formata";

import { 
    Container, 
    Title, 
    Item, 
    ItemPress,
    ItemName, 
    ItemDescription, 
    ColumLeft,
    ColumRight ,
    ViewLoading,
    TextLoading,
    ViewError,
    TextError
} from './style';

interface acao {
    id: number;
    nome: string;
    percentual: number
} 

interface Investimento {
    nome: string;
    objetivo: string;
    saldoTotal: number;
    indicadorCarencia: string;
    acoes: acao[]
} 

interface ListaInvestimentos {
    listaInvestimentos: Investimento[]
}

const ItemInvestimento = props => {        

    const { navigate } = props.navigation;
    
    const handlePress = (inv: Investimento) => {
        navigate('Resgate', inv);
    }

    const Columns = props => (
        <>
            <ColumLeft>
                <ItemName>{ props.inv.nome }</ItemName>
                <ItemDescription>{ props.inv.objetivo }</ItemDescription>
            </ColumLeft>
            <ColumRight>{ formatNumber(props.inv.saldoTotal) }</ColumRight>
        </>
    );

    if(props.inv.indicadorCarencia == 'S'){
        return (            
            <Item> 
                <Columns {...props} />
            </Item>            
        );
    }else{
        return (
            <ItemPress onPress={() => handlePress(props.inv)}> 
                <Columns {...props} />
            </ItemPress>
        );
    }
};

const Investimento: React.FC = (props) => {
    
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(false);
    const [data, setData] = useState<ListaInvestimentos>();

    const loadApi = async() => {                     
        try {                   
            let { data } = await api.get('ca4ec77d-b941-4477-8a7f-95d4daf7a653');            
            setData(data.response.data);                                    
        } catch (error) {   
            setError(true);            
        }         
        setLoading(false);
    };

    useEffect(() => {        
        loadApi();
    }, []);

    if(loading) {
        return (
            <ViewLoading>
                <TextLoading>Carregando...</TextLoading>
            </ViewLoading>
        );
    } else {
        if(error){
            return(
                <ViewError>       
                    <TextError>ERRO DE CONEX√ÉO</TextError>                                        
                </ViewError>
            );
        }else{
            return (
                <Container>       
                    <Title>                    
                        <ColumLeft>
                            <ItemName style={{color:'#898989'}}>INVESTIMENTOS</ItemName>
                        </ColumLeft>
                        <ColumRight style={{color:'#898989'}}>R$</ColumRight>
                    </Title>                     
                    {data.listaInvestimentos.map((e, index) => 
                        <ItemInvestimento {...props} key={index} inv={e} />
                    )}
                </Container>
            );
        }
    }
}

export default Investimento;