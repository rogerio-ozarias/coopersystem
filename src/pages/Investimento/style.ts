import styled from 'styled-components/native'; 

export const Container = styled.View`
    display: flex;
    background: #EFEFEF;
    width: 100%;
    height: 100%;    
`;

export const Title = styled.View`	
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  color: #656565;    
  padding: 20px;
`;

export const Item = styled.View`  
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  background: #FFFFFF;
  color: #252525;
  padding: 20px;
  margin-bottom: 2px;  
`;
export const ItemPress = styled.TouchableOpacity`  
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  background: #FFFFFF;
  color: #252525;
  padding: 20px;
  margin-bottom: 2px;  
`;

export const ItemName = styled.Text`
  color: #565656
`;

export const ItemDescription = styled.Text`
  color: #888888
`;

export const ColumLeft = styled.View``;

export const ColumRight = styled.Text`
  color: #565656
`;

export const ViewLoading = styled.View`
  display: flex;
  height: 100%;
  align-items: center;
  justify-content: center;
`;

export const TextLoading = styled.Text`    
  font-weight: bold;
`;

export const ViewError = styled.View`
  display: flex;
  height: 100%;
  align-items: center;
  justify-content: center;
`;

export const TextError = styled.Text`    
  font-weight: bold;
  color: #F00;
`;