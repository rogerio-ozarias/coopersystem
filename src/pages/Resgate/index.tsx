import React, { useState, useReducer, useEffect } from 'react';
import { ScrollView, Modal, Alert } from 'react-native';
import { useRoute, useNavigation, Link } from '@react-navigation/native';
import { TextInputMask } from 'react-native-masked-text';
import { formatNumber } from "../../helpers/Formata";
import Button from '../components/Button';


import { 
    Container, 
    Title, 
    ItemName,   
    ColumLeft,
    RowBox,    
    RowBoxLabel,
    RowBoxInput,
    RowBoxInputInfo,
    ColBox,
    ColBoxTitle,
    BoxAcao,
    ViewModal,
    ModalContainer,
    ModalContainerValido,
    ViewModalContent,
    ViewModalTitle,
    ViewModalSubTitle,
    ViewModalText
} from './style';

interface AcaoInterface {
    id: number;
    nome: string;
    percentual: number,
    inputInfo: string,
    valorResgatar: number
} 

interface InvestimentoInterface {
    nome: string;
    objetivo: string;
    saldoTotal: number;
    indicadorCarencia: string;
    acoes: AcaoInterface[]
} 

const ValorResgate = (props) => {
    const [valorResgatar, setValorResgatar] = useState('0');  
    const [infoResgate, setInfoResgate] = useState('');    
    const [labelColor, setLabelColor] = useState('#888');    
    const { acao, saldoTotal, setValorTotal } = props;
    const valorDispoinivel = saldoTotal*(acao.percentual/100);        
    
    return (
        <BoxAcao>
            <RowBox>
                <ColBox style={{color:'#000'}}>Ação</ColBox>
                <ColBox>{ acao.nome }</ColBox>
            </RowBox>
            <RowBox>
                <ColBox style={{color:'#000'}}>Saldo acumulado</ColBox>
                <ColBox>R$ { formatNumber(valorDispoinivel) }</ColBox>
            </RowBox>  
            <RowBoxLabel style={{color:labelColor}}>Valor a resgatar</RowBoxLabel>
            <RowBoxInput>
                <TextInputMask { ...props } 
                    type={'money'}
                    options={{
                        precision: 2,
                        separator: ',',
                        delimiter: '.',
                        unit: 'R$ ',
                        suffixUnit: ''                
                    }}
                    value={ valorResgatar }
                    includeRawValueInChangeText={true}
                    onChangeText={(text, rawText) => {
                         
                        let valorDBResgate = parseFloat(rawText);                                                
                        acao.inputInfo =  (valorDBResgate > valorDispoinivel) ? 
                            setInfoResgate(`Valor não pode ser maior que R$ ${formatNumber(valorDispoinivel)}`)
                        :   setInfoResgate('');                        

                        setValorTotal({
                            type: 'add',
                            item:{
                                id: acao.id, 
                                nome: acao.nome,
                                valor: valorDBResgate,
                                valorDisponivel: valorDispoinivel
                            }                            
                        });                        
                    }}                                
                    onFocus={(e) => {
                        setLabelColor('#23569C');
                    }}  
                    onBlur={() => {
                        setLabelColor('#888');
                    }}                         
                    keyboardType="numeric"  />
            </RowBoxInput>
            <RowBoxInputInfo>
                { infoResgate }
            </RowBoxInputInfo>
           
        </BoxAcao>
    );
}

const Resgate: React.FC = (props) => {
    
    const route = useRoute();
    const routeParams = route.params as InvestimentoInterface;
    const { saldoTotal } = routeParams;          
    const [valorTotal, setValorTotal] = useState(0.00);    
    const [validacao, setValidacao] = useState(false);       
    const [modalVisible, setModalVisible] = useState(false);
    const [modalTitle, setModalTitle] = useState('');
    const [modalSubTitle, setModalSubTitle] = useState('');
    const [modalText, setModalText] = useState([]);    
    const navigation = useNavigation();    

    const initialState = [];
    const reducer = (resgateState, action ) => {
        switch(action.type){
            case 'add':                   
                let newState = resgateState.filter(e => e.id !== action.item.id);
                newState = [...newState, action.item];                  
                let somaValor = 0;
                newState.map(e => {
                    somaValor+= (!isNaN(e.valor)) ? e.valor : 0;                    
                })
                setValorTotal(somaValor);                                
                return newState;                
            break;            
            default:
                return resgateState
        }
    };

    const [resgateState, dispatch] = useReducer(reducer, initialState);
    
    const handlePres = () => {
        let erro = [];
        let valido = false;
        // verifica se for informado valor a resgatar
        if(valorTotal > 0){                        
            // faz a validação
            resgateState.map(e => {
                if(e.valor >  e.valorDisponivel){                    
                    erro.push(`${e.nome}: Valor máximo de R$ ${formatNumber(e.valorDisponivel)}`);
                }
            })        
            if(erro.length){
                setModalTitle('DADOS INVÁLIDOS');
                setModalSubTitle('Você preencheu um ou mais campos acima do permitido:');
                setModalText(erro);
            }else{
                valido = true;
                setModalTitle('RESGATE EFETUADO!');
                setModalSubTitle('O valor solicitado estará em sua conta em até 5 dias uteis!');                
                setModalText(erro);
            }
        }else{
            setModalTitle('DADOS INVÁLIDOS');
            setModalSubTitle('Você não informou nenhum valor de resgate.');            
            setModalText(erro);
        }    
        
        setValidacao(valido);                   
        setModalVisible(true);                                      
             
    } 
    
    const ModalBox = (props) => {
        return (
            <Modal animationType='fade' visible={props.modalVisible} transparent={true}>
                <ModalContainer>
                    <ViewModal>   
                        <ViewModalContent>                    
                            <ViewModalTitle>{modalTitle}</ViewModalTitle>
                            <ViewModalSubTitle>{modalSubTitle}</ViewModalSubTitle>
                            { modalText.map((e, index) =>  <ViewModalText key={index}>{ e }</ViewModalText>) }
                        </ViewModalContent>
                        <Button title="FECHAR" onPress={props.onClose()} />   
                    </ViewModal>
                </ModalContainer>
            </Modal>
        );
    }
    
    return (
        <Container>                            
            
            <ScrollView>
                <Title>                    
                    <ColumLeft>
                        <ItemName style={{color:'#898989'}}>DADOS DO INVESTIMENTO</ItemName>
                    </ColumLeft>
                </Title>                 

                <RowBox>
                    <ColBox style={{ color:'#000'}}>Nome</ColBox>
                    <ColBox>{ routeParams.nome }</ColBox>
                </RowBox>
                <RowBox>
                    <ColBox style={{ color:'#000'}}>Saldo total dispoinivel</ColBox>
                    <ColBox>R$ { formatNumber(routeParams.saldoTotal) }</ColBox>
                </RowBox>       

                <Title>                    
                    <ColumLeft>
                        <ItemName style={{color:'#898989'}}>RESGATE DO SEU JEITO</ItemName>
                    </ColumLeft>
                </Title>        
                
                {routeParams.acoes.map((e, index) => 
                    <ValorResgate 
                        key={index} 
                        setValorTotal={dispatch} 
                        saldoTotal={saldoTotal} 
                        acao={e} placeholder='Digite o valor' /> 
                )}

                <RowBox>
                    <ColBoxTitle>Valor total a resgatar</ColBoxTitle>
                    <ColBox>R$ { formatNumber(valorTotal) }</ColBox>
                </RowBox>                  
                
                <Button title="CONFIRMAR RESGATE" onPress={handlePres}/>    

                {
                    // se tiver valido
                    validacao ?
                        <Modal animationType='fade' visible={modalVisible} transparent={false} >
                            <ModalContainerValido>
                                <ViewModal>   
                                    <ViewModalContent>                    
                                        <ViewModalTitle>{modalTitle}</ViewModalTitle>
                                        <ViewModalSubTitle>{modalSubTitle}</ViewModalSubTitle>
                                        { modalText.map((e, index) =>  <ViewModalText key={index}>{ e }</ViewModalText>) }
                                    </ViewModalContent>
                                    
                                    <Button title="NOVO RESGATE" onPress={() => {                                                                         
                                        navigation.goBack();                     
                                        setModalVisible(false);                  
                                    }} /> 
                                    
                                </ViewModal>
                            </ModalContainerValido>
                        </Modal>

                    :
                        // se tiver invalido
                        <Modal animationType='fade' visible={modalVisible} transparent={true} >
                            <ModalContainer>
                                <ViewModal>   
                                    <ViewModalContent>                    
                                        <ViewModalTitle>{modalTitle}</ViewModalTitle>
                                        <ViewModalSubTitle>{modalSubTitle}</ViewModalSubTitle>
                                        { modalText.map((e, index) =>  <ViewModalText key={index}>{ e }</ViewModalText>) }
                                    </ViewModalContent>
                                   
                                    <Button title="CORRIGIR" onPress={() => {
                                        setModalVisible(false);                                                                        
                                    }} />   
                                    
                                </ViewModal>
                            </ModalContainer>
                        </Modal>
                }                                 
                
            </ScrollView>
        </Container>
    );
}

export default Resgate;