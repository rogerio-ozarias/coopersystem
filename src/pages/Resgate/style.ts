import styled from 'styled-components/native'; 

export const Container = styled.View`
    display: flex;
    background: #EFEFEF;
    width: 100%;
    height: 100%;    
`;

export const Title = styled.View`	
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  color: #656565;    
  padding: 20px;
`;

export const Item = styled.View`  
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  background: #FFFFFF;
  color: #252525;
  padding: 20px;
  margin-bottom: 2px;  
`;

export const ItemName = styled.Text`
  color: #565656;
`;

export const ItemDescription = styled.Text`
  color: #888888;
`;

export const ColumLeft = styled.View``;

export const ColumRight = styled.Text`
  color: #565656;
`;

export const RowBox = styled.View`  
  display: flex;
  width: 100%;
  flex-direction: row;
  justify-content: space-between;
  background: #FFFFFF;
  color: #252525;
  padding: 10px 20px;
  border: solid 1px #EFEFEF;  
`;


export const RowForm = styled.View`
  border: solid 2px #EFEFEF;
`;
export const RowBoxLabel = styled.Text`  
  width: 100%;  
  background: #FFFFFF;
  color: #656565;  
  padding-left:20px;
`;
export const RowBoxInput = styled.View`  
  width: 100%;  
  background: #FFFFFF;
  color: #252525;
  padding-left:20px;
`;
export const RowBoxInputInfo = styled.Text`
  width: 100%;  
  color: #FF0000;
  background: #FFFFFF;
  padding-left:20px;
`;



export const ColBox = styled.Text`
  display: flex;
  flex-direction: row;
`;

export const ColBoxTitle = styled.Text`
  display: flex;
  flex-direction: row;
  color: #000000;
  font-weight: bold
`;

export const BoxAcao = styled.View`
  margin-bottom: 10px;
`;

export const ModalContainer = styled.View`
  background-color: rgba(0,0,0,0.7);
  flex: 1;
`;

export const ModalContainerValido = styled.View`
  background-color: rgba(150, 150, 150, 6);
  flex: 1;
`;

export const ViewModal = styled.View`  
  width: 90%;    
  margin: auto;  
  background: #FFFFFF;
`;

export const ViewModalContent = styled.View`  
  padding: 20px;
  background: #FFFFFF;
`;

export const ViewModalTitle = styled.Text`
  color: #23569C;  
  margin-bottom: 10px;
  font-weight: bold;
  font-size: 22px;
  text-align: center
`;

export const ViewModalSubTitle = styled.Text`
  width: 100%;
  color: #656565;  
  font-size: 16px;   
  margin-bottom: 30px;
`;

export const ViewModalText = styled.Text`
  width: 100%;
  color: #656565;  
  font-size: 16px;      
`;