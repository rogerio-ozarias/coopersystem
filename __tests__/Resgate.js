import React from 'react';
import renderer from 'react-test-renderer';
import Investimento from "../src/pages/Investimento";

test('renders correctly', () => {
  const tree = renderer.create(<Investimento />).toJSON();
  expect(tree).toMatchSnapshot();
});
/*
import { renderer, screen} from 'react-test-renderer';
import Invesvestimento from "../src/pages/Investimento";
import Resgate from "../src/pages/Resgate";

// https://www.youtube.com/watch?v=pbwXsjVEMqg
// https://jestjs.io/pt-BR/docs/tutorial-react-native
describe('Clicar em confirmar com mais de um campos a resgatar com valor invalido', () => {
    test('Clicar no investimento I', () => {
        // testar se abru a tela de resgate após o clique
        const { investimento } = renderer(<Investimento />)
        return true;
    });

    test('Digitar um valor acima do disponível na primeira ação.', () => {
        // Ex: BBAS3 tem 11 mil de saldo acumulado, digitar 12 mil no campo valor a resgatar
        return true;
    });

    test('Digitar um valor abaixo do disponível na segunda ação.', () => {
        // Ex: VALE3 tem 8 mil de saldo acumulado, digitar 9 mil no campo valor a resgatar
        return true;
    });

    test('Clicar em confirmar', () => {
        
        // Resultado esperado: Deve aparecer um modal alertando 
        // que foi digitado um ou mais valor acima do permitido, 
        // e exibir quais ações estão com erro.
        
        return true;
    });
})

describe('Clicar em confirmar com todos os campos com valor validos', () => {
    test('Clicar no investimento I', () => {
        // testar se abril a tela de resgate
        return true;
    });

    test('Digitar um valor abaixo ou igual ao disponível na primeira e segunda ação.', () => {
        // Ex: BBAS3 tem 11 mil de saldo acumulado, digitar 11 mil no campo valor a resgatar 
        // Ex: VALE3 tem 8 mil de saldo acumulado, digitar 2 mil no campo valor a resgatar
        return true;
    });

    test('Digitar um valor abaixo ou igual ao disponível na terceira ação.', () => {
        return true;
    });

    test('Clicar em confirmar', () => {
        
        // Resultado esperado: Deve aparecer um modal com a mensagem que o 
        // resgate foi efetuado com sucesso, e quando clicar em novo resgate, 
        // voltar para tela inicial.
   
        return true;
    });
})*/