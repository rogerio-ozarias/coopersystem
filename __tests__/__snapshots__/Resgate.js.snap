// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontWeight": "bold",
        },
      ]
    }
  >
    Carregando...
  </Text>
</View>
`;
